truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedDispatchNode.java

```diff
@@ -63,20 +63,32 @@ public CachedDispatchNode(
         this.next = next;
     }
 
+    protected boolean sendingFrames() {
+        return sendsFrame || readCaller != null;
+    }
+
     public void replaceSendingFrame() {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert readCaller == null;
         if (!sendsFrame) {
-            sendsFrame = true;
+            replaceSendingFrame(true, null);
         }
     }
 
     public void replaceSendingCallerFrame(CallerFrameAccess access) {
-        CompilerDirectives.transferToInterpreterAndInvalidate();
+        assert sendsFrame == false;
         if (readCaller == null) {
-            readCaller = new ReadCallerFrameNode(access);
+            replaceSendingFrame(false, new ReadCallerFrameNode(access));
         }
     }
 
+    private void replaceSendingFrame(boolean sendsFrame, ReadCallerFrameNode readCaller) {
+        CachedDispatchNode copy = (CachedDispatchNode) copy();
```

Comment:
I am not sure we need to copy here.
What we need is to invalidate the encapsulating CallTarget Assumption.
Would `replace(this)` work?
It's a bit hacky but I am not aware of a better way to invalidate that Assumption.